File: shared_tracker/extension/autofill-profile.js
"use strict";
const form = document.querySelector('#autofill-form'), status = document.querySelector('#status');
const saveButton = document.querySelector('#save-profile'), loadStatus = document.querySelector('#load-status');
const retryButton = document.querySelector('#retry-load');
const PROFILE_URL = 'http://127.0.0.1:8767/api/autofill-profile';
const scalarFields = {prefix:'Prefix',full_name:'Full legal name',first_name:'First / given name',last_name:'Last / family name or surname',email:'Email',phone:'Phone number',phone_device_type:'Phone device type',country:'Country of residence',street_name:'Street name / address line 1',additional_address:'Additional address / address line 2',city:'City',postal_code:'Postal code',linkedin_url:'LinkedIn website',website_url:'Website 1',website_url_2:'Website 2',gender:'Gender',date_of_birth:'Date of birth',country_of_birth:'Country / territory of birth',ethnicity:'Race / ethnicity',religion:'Religion',nationality:'Primary nationality',additional_nationalities:'Additional nationalities (comma separated)',citizenship:'Citizenship / Singapore right-to-work status',years_work_experience:'Full-time working experience in years (excluding internships and part-time work)',expected_salary:'Expected monthly base salary',current_monthly_base_salary:'Current monthly base salary',current_annual_bonus:'Current annual bonus',current_variable_compensation:'Current variable compensation / income',salary_currency:'Salary currency',notice_period:'Notice required before leaving current employer',earliest_available_start_date:'Earliest available start date',work_authorized:'Legally authorised to work in job country',requires_sponsorship:'Visa / work-authorisation sponsorship required',holds_required_credentials:'Hold job-required certifications or clearance',willing_to_travel:'Willing and able to travel',preferred_office_location:'Preferred office location',professional_memberships:'Professional, social or sports memberships',financial_interest:'Financial interest disclosure',outside_employment_or_business:'Outside employment / business activity disclosure',disciplinary_history:'Dismissal / disciplinary history',criminal_record:'Criminal / regulatory proceedings',related_to_employer:'Related to an employee / partner (details or N/A)',relatives_at_employer:'Relatives employed by the organisation (details or N/A)',referral_source:'How did you hear about this role?',future_recruitment_consent:'Consent to recruitment for other positions',communications_consent:'Consent to job-related event communications',declaration_acknowledgement:'Accept application declaration',summary:'Professional summary',skills:'Resume key skills (comma separated)',certifications:'Certifications'};
const resumeScalarKeys = new Set(['years_work_experience', 'summary', 'skills', 'certifications']);
const historyFields = {education:{institution:'Institution',qualification:'Qualification',field_of_study:'Field of study',start_date:'Start date',end_date:'Completion date',country:'Country',grade:'Grade / result'},employment:{employer:'Employer',job_title:'Job title',start_date:'Start date',end_date:'End date',current:'Currently working here',country:'Country',description:'Responsibilities and achievements',reason_for_leaving:'Reason for leaving',salary:'Salary in this job (optional)'}};
const yesNo = ['','Yes','No'];
const choiceFields = {prefix:['','Dr.','Miss','Mr.','Mrs.','Ms.','Prof.'],phone_device_type:['','Mobile','Landline'],work_authorized:yesNo,requires_sponsorship:yesNo,holds_required_credentials:yesNo,willing_to_travel:yesNo,future_recruitment_consent:yesNo,communications_consent:yesNo,financial_interest:[...yesNo,'N/A'],outside_employment_or_business:[...yesNo,'N/A'],disciplinary_history:[...yesNo,'N/A'],criminal_record:[...yesNo,'N/A'],declaration_acknowledgement:[...yesNo,'N/A']};
let loaded = null, dirty = false, loading = null, saving = false;
async function fetchLocalProfile(method = 'GET', value) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 20000);
  try {
    const response = await fetch(PROFILE_URL, {
      method, cache: 'no-store', signal: controller.signal,
      headers: {'X-Job-Assistant-Extension': chrome.runtime.id, ...(method === 'POST' ? {'Content-Type': 'application/json'} : {})},
      ...(method === 'POST' ? {body: JSON.stringify(value)} : {}),
    });
    const data = await response.json();
    if (!response.ok || !data?.ok) throw new Error(data?.error || 'The local assistant could not load your profile.');
    return data;
  } finally { clearTimeout(timeout); }
}
async function profileRequest(method = 'GET', value) {
  if (method === 'POST') return fetchLocalProfile(method, value);
  try { return await fetchLocalProfile(); }
  catch (error) {
    // Only connection failures need the native helper. Never retry a save:
    // a timed-out POST might already have committed the user's answers.
    if (!(error instanceof TypeError || error.name === 'AbortError')) throw error;
    try {
      const viaWorker = await chrome.runtime.sendMessage({type:'GET_SF_PROFILE'});
      if (viaWorker?.ok) return viaWorker;
    } catch (_) { /* The direct retry below reports the final failure. */ }
    return fetchLocalProfile();
  }
}
function input(label, value, key) {
  const wrapper = document.createElement('label'); wrapper.textContent = label;
  const field = document.createElement(choiceFields[key] ? 'select' : /summary|skills|certifications|description|answer/.test(key) ? 'textarea' : 'input');
  if (choiceFields[key]) for (const optionValue of choiceFields[key]) {
    const option = document.createElement('option'); option.value = optionValue; option.textContent = optionValue || 'Not set'; field.appendChild(option);
  }
  if (key === 'current') { field.type = 'checkbox'; field.checked = value === true; wrapper.className = 'check'; }
  else { field.value = value ?? ''; if (field.tagName === 'INPUT') field.type = ['earliest_available_start_date','date_of_birth'].includes(key) ? 'date' : 'text'; }
  field.dataset.key = key; wrapper.appendChild(field);
  if (field.tagName === 'TEXTAREA') wrapper.className = 'wide';
  return wrapper;
}
function renderHistory(section, entries) {
  const host = document.getElementById(section); host.replaceChildren();
  entries.forEach((entry, index) => {
    const card = document.createElement('div'); card.className = 'entry';
    const title = document.createElement('h3'); title.textContent = (section === 'education' ? 'Education ' : 'Experience ') + (index + 1);
    const grid = document.createElement('div'); grid.className = 'grid';
    for (const [key,label] of Object.entries(historyFields[section])) grid.appendChild(input(label, entry[key], key));
    const remove = document.createElement('button'); remove.type = 'button'; remove.textContent = 'Remove entry'; remove.onclick = () => { card.remove(); dirty = true; };
    card.append(title, grid, remove); host.appendChild(card);
  });
}
function history(section) {
  return [...document.querySelectorAll('#' + section + ' .entry')].map(card => Object.fromEntries([...card.querySelectorAll('[data-key]')].map(field => [field.dataset.key,field.type === 'checkbox' ? field.checked : field.value.trim()])));
}
function sameHistory(section, left, right) {
  const normalize = entries => entries.map(entry => Object.keys(historyFields[section]).map(key => key === 'current' ? entry[key] === true : String(entry[key] ?? '').trim()));
  return JSON.stringify(normalize(left)) === JSON.stringify(normalize(right));
}
function addAnswer(value = {}) {
  const row = document.createElement('div'); row.className = 'entry grid';
  row.append(input('Exact question', value.question, 'question'), input('Your answer', value.answer, 'answer'));
  const remove = document.createElement('button'); remove.type = 'button'; remove.textContent = 'Remove answer'; remove.onclick = () => { row.remove(); dirty = true; };
  row.appendChild(remove); document.querySelector('#custom-answers').appendChild(row);
}
function render(data) {
  if (!data?.ok || !data.storage_revision || !data.profile) throw new Error('The local assistant returned an incomplete profile. Restart it and reload this extension.');
  loaded = data; dirty = false;
  document.querySelector('#enabled').checked = data.enabled;
  document.querySelector('#source').textContent = data.source ? 'Defaults extracted automatically from: ' + data.source.name : 'No source resume found.';
  document.querySelector('#warnings').textContent = data.warnings.join(' ');
  const manualHost = document.querySelector('#contact-fields'), resumeHost = document.querySelector('#resume-fields');
  manualHost.replaceChildren(); resumeHost.replaceChildren();
  for (const [key,label] of Object.entries(scalarFields)) {
    const wrapper = input(label, data.profile[key], key);
    const tools = document.createElement('div'); tools.className = 'field-tools';
    const fromResume = Boolean(data.defaults[key]);
    const hint = document.createElement('span'); hint.textContent = key in data.overrides ? 'Your saved value' : fromResume ? 'From newest resume' : 'Enter and save manually';
    const reset = document.createElement('button'); reset.type = 'button'; reset.textContent = fromResume ? 'Use resume default' : 'Clear saved value';
    reset.onclick = () => { wrapper.querySelector('[data-key]').value = fromResume ? loaded.defaults[key] || '' : ''; delete loaded.overrides[key]; hint.textContent = fromResume ? 'From newest resume' : 'Enter and save manually'; dirty = true; };
    tools.append(hint,reset); wrapper.appendChild(tools); (resumeScalarKeys.has(key) ? resumeHost : manualHost).appendChild(wrapper);
  }
  for (const section of Object.keys(historyFields)) renderHistory(section, data.profile[section] || []);
  document.querySelector('#custom-answers').replaceChildren(); data.custom_answers.forEach(addAnswer);
  form.hidden = false;
  saveButton.disabled = false;
  retryButton.hidden = true;
  loadStatus.textContent = 'Saved answers loaded from this computer.';
}
async function load(force = false) {
  if (saving) return;
  if (dirty && !force) return;
  if (dirty && force && !confirm('Discard unsaved edits and read the newest resume?')) return;
  if (loading) return loading;
  loading = (async () => {
    const data = await Promise.race([
      profileRequest(),
      new Promise((_, reject) => setTimeout(() => reject(new Error('The local assistant did not respond within 45 seconds. Run Setup Job Assistant Suite.bat, then reload this extension.')), 45000)),
    ]);
    if (!data?.ok) throw new Error(data?.error || 'Start the shared Chrome helper.');
    if (dirty && !force) return;
    if (force || data.revision !== loaded?.revision) { render(data); status.textContent = 'Newest resume defaults loaded. Your saved corrections are retained.'; }
  })();
  try { await loading; }
  catch (error) {
    if (!loaded) {
      form.hidden = true;
      saveButton.disabled = true;
      retryButton.hidden = false;
      document.querySelector('#source').textContent = 'Saved answers could not be loaded.';
      loadStatus.textContent = error.message + ' Do not re-enter or save details until this page loads.';
    }
    throw error;
  } finally { loading = null; }
}
form.addEventListener('input', () => { dirty = true; });
form.addEventListener('submit', async event => {
  event.preventDefault();
  if (!loaded?.ok || !loaded.storage_revision || saving) {
    status.textContent = 'Wait for your saved answers to load before saving.';
    return;
  }
  const overrides = {...loaded.overrides};
  for (const field of document.querySelectorAll('#contact-fields [data-key], #resume-fields [data-key]')) {
    const key = field.dataset.key, value = field.value.trim();
    if (value !== (loaded.defaults[key] || '')) overrides[key] = value; else delete overrides[key];
  }
  for (const section of Object.keys(historyFields)) {
    const entries = history(section);
    if (!sameHistory(section, entries, loaded.defaults[section] || [])) overrides[section] = entries; else delete overrides[section];
  }
  const custom_answers = [...document.querySelectorAll('#custom-answers .entry')].map(row => ({question:row.querySelector('[data-key="question"]').value.trim(),answer:row.querySelector('[data-key="answer"]').value.trim()})).filter(item => item.question);
  saving = true; saveButton.disabled = true;
  try {
    const data = await profileRequest('POST', {enabled:document.querySelector('#enabled').checked,overrides,custom_answers,base_storage_revision:loaded.storage_revision});
    if (!data?.ok) throw new Error(data?.error || 'Save failed.');
    render(data); status.textContent = 'Saved. The unified extension uses these defaults.';
  } catch(error) { status.textContent = error.message; }
  finally { saving = false; saveButton.disabled = false; }
});
document.querySelectorAll('[data-add]').forEach(button => button.onclick = () => { const section = button.dataset.add; renderHistory(section, [...history(section), {}]); dirty = true; });
document.querySelectorAll('[data-reset]').forEach(button => button.onclick = () => { const section = button.dataset.reset; renderHistory(section, loaded.defaults[section] || []); delete loaded.overrides[section]; dirty = true; });
document.querySelector('#add-answer').onclick = () => { addAnswer(); dirty = true; };
document.querySelector('#reload').onclick = () => load(true).catch(error => { status.textContent = error.message; });
retryButton.onclick = () => load(true).catch(error => { status.textContent = error.message; });
setInterval(() => load().catch(error => { status.textContent = error.message; }), 10000);
status.textContent = 'Connecting to the local assistant…';
load().catch(error => { status.textContent = error.message; });

File: shared_tracker/autofill_profile.py
"""One private, editable profile shared by the three Chrome extensions."""
from __future__ import annotations

import hashlib
import json
import threading
from datetime import date
from pathlib import Path
from uuid import uuid4
from .schema import LOCAL_DATA, WORKSPACE
from .resume_profile import newest_resume, read_resume, parse_resume

PROFILE_PATH = LOCAL_DATA / 'private/autofill-profile.local.json'
# Only facts explicitly extracted from the newest resume become defaults.
# Application choices, name order, salary and dates that need inference stay manual.
RESUME_DEFAULT_FIELDS = frozenset({
    'full_name', 'email', 'phone', 'linkedin_url', 'website_url', 'website_url_2',
    'street_name', 'additional_address', 'city', 'postal_code', 'country',
    'summary', 'skills', 'certifications', 'education', 'employment',
})
TEXT_FIELDS = ('prefix','full_name','first_name','last_name','email','phone','phone_device_type','country','city',
               'street_name','additional_address','address_line1','postal_code','linkedin_url','website_url','website_url_2',
               'expected_salary','salary_currency','notice_period','gender','date_of_birth','country_of_birth','ethnicity',
               'religion','nationality','additional_nationalities',
               'citizenship','work_authorized','requires_sponsorship','holds_required_credentials','willing_to_travel',
               'earliest_available_start_date','related_to_employer','future_recruitment_consent','communications_consent',
               'current_monthly_base_salary','current_annual_bonus','current_variable_compensation','preferred_office_location',
               'professional_memberships','financial_interest','outside_employment_or_business','disciplinary_history',
               'criminal_record','relatives_at_employer','referral_source','declaration_acknowledgement',
               'summary','skills','certifications')
_lock = threading.RLock()
_cache = {}


def allowed_extension(origin: str) -> bool:
    manifest = WORKSPACE / 'chrome-helper/generated/com.job_application_assistant.launcher.json'
    try: allowed = json.loads(manifest.read_text(encoding='utf-8'))['allowed_origins']
    except (OSError, ValueError, KeyError): return False
    return origin.rstrip('/') + '/' in allowed and origin.startswith('chrome-extension://')


def allowed_extension_request(origin: str, extension_id: str) -> bool:
    """Accept privileged extension fetches that omit Origin, not web origins."""
    if origin:
        return allowed_extension(origin)
    if not extension_id or len(extension_id) != 32 or any(letter not in 'abcdefghijklmnop' for letter in extension_id):
        return False
    return allowed_extension('chrome-extension://' + extension_id)


def validate_saved(value: dict) -> dict:
    if not isinstance(value, dict) or set(value) - {'enabled', 'overrides', 'custom_answers'}:
        raise ValueError('Use the autofill profile form to edit these settings.')
    if not isinstance(value.get('enabled', True), bool): raise ValueError('Enabled must be true or false.')
    overrides = value.get('overrides', {})
    if not isinstance(overrides, dict) or set(overrides) - set(TEXT_FIELDS) - {'education','employment'}:
        raise ValueError('Unknown profile fields.')
    for key, item in overrides.items():
        if key in TEXT_FIELDS and (not isinstance(item, str) or len(item) > 20000): raise ValueError('Profile text is too long or invalid.')
        if key == 'prefix' and item not in {'', 'Dr.', 'Miss', 'Mr.', 'Mrs.', 'Ms.', 'Prof.'}:
            raise ValueError('Choose a supported prefix from the profile form.')
        if key == 'phone_device_type' and item not in {'', 'Mobile', 'Landline'}:
            raise ValueError('Phone device type must be Mobile or Landline.')
        if key in {'work_authorized','requires_sponsorship','holds_required_credentials','willing_to_travel',
                   'future_recruitment_consent','communications_consent'} and item not in {'', 'Yes', 'No'}:
            raise ValueError('Choose Yes, No or Not set for this application answer.')
        if key in {'financial_interest','outside_employment_or_business','disciplinary_history','criminal_record',
                   'declaration_acknowledgement'} and item not in {'', 'Yes', 'No', 'N/A'}:
            raise ValueError('Choose Yes, No, N/A or Not set for this application answer.')
        if key in {'earliest_available_start_date', 'date_of_birth'} and item:
            try: date.fromisoformat(item)
            except ValueError as exc: raise ValueError(f'{key.replace("_", " ").title()} must use YYYY-MM-DD.') from exc
        if key in {'education','employment'}:
            fields = {'institution','qualification','field_of_study','start_date','end_date','country','grade'} if key == 'education' else {'employer','job_title','start_date','end_date','current','description','country','reason_for_leaving','salary'}
            if not isinstance(item, list) or len(item) > 30: raise ValueError('Use at most 30 history entries.')
            for entry in item:
                if not isinstance(entry, dict) or set(entry) - fields: raise ValueError('Invalid history fields.')
                for field, val in entry.items():
                    if field == 'current':
                        if not isinstance(val, bool): raise ValueError('Current employment must be true or false.')
                    elif not isinstance(val, str) or len(val) > 20000: raise ValueError('Invalid history value.')
    answers = value.get('custom_answers', [])
    if not isinstance(answers, list) or len(answers) > 100: raise ValueError('Use at most 100 saved answers.')
    for answer in answers:
        if not isinstance(answer, dict) or set(answer) != {'question','answer'} or not all(isinstance(v, str) and len(v) <= 20000 for v in answer.values()):
            raise ValueError('Saved answers need a question and answer.')
    return {'enabled': value.get('enabled', True), 'overrides': overrides, 'custom_answers': answers}


def get_profile(path: Path = PROFILE_PATH, folders: list[Path] | None = None) -> dict:
    with _lock:
        saved = validate_saved(json.loads(path.read_text(encoding='utf-8'))) if path.exists() else validate_saved({})
        resume = newest_resume(folders if folders is not None else [LOCAL_DATA / 'resumes'])
        defaults, warnings, source = {}, [], None
        if resume:
            signature = (str(resume.resolve()), resume.stat().st_mtime_ns, resume.stat().st_size)
            if signature not in _cache:
                try:
                    text = read_resume(resume)
                    if not text.strip(): raise ValueError('This resume has no readable text (it may need OCR).')
                    _cache.clear()
                    _cache[signature] = parse_resume(text)
                except Exception as exc:
                    _cache[signature] = ({}, ['The newest resume could not be parsed: ' + str(exc)])
            extracted, warnings = _cache[signature]
            defaults = {key: value for key, value in extracted.items() if key in RESUME_DEFAULT_FIELDS}
            source = {'name': resume.name, 'modified_ns': signature[1]}
        else: warnings = ['Add a source PDF or DOCX resume to local-data/resumes.']
        effective = {**{key: '' for key in TEXT_FIELDS}, 'education': [], 'employment': [], **defaults, **saved['overrides']}
        if effective['first_name'] and effective['last_name']:
            warnings = [warning for warning in warnings if not warning.startswith('Given/family name order')]
        result = {**saved, 'defaults': defaults, 'profile': effective, 'source': source, 'warnings': warnings,
                  'storage_revision': _storage_revision(saved)}
        result['revision'] = hashlib.sha256(json.dumps(result, sort_keys=True).encode()).hexdigest()
        return result


def _storage_revision(saved: dict) -> str:
    """Track edits to saved answers independently of resume-derived defaults."""
    return hashlib.sha256(json.dumps(saved, sort_keys=True, ensure_ascii=False).encode('utf-8')).hexdigest()


def save_profile(value: dict, path: Path = PROFILE_PATH) -> dict:
    if not isinstance(value, dict):
        raise ValueError('Reload the autofill profile before saving.')
    base_revision = value.get('base_storage_revision')
    if not isinstance(base_revision, str) or len(base_revision) != 64:
        raise ValueError('Reload the autofill profile before saving. No changes were saved.')
    saved = validate_saved({key: item for key, item in value.items() if key != 'base_storage_revision'})
    with _lock:
        current = validate_saved(json.loads(path.read_text(encoding='utf-8'))) if path.exists() else validate_saved({})
        if _storage_revision(current) != base_revision:
            raise ValueError('The autofill profile changed in another tab. Reload it before saving. No changes were saved.')
        path.parent.mkdir(parents=True, exist_ok=True)
        temporary = path.with_name('.profile-' + uuid4().hex + '.tmp')
        backup_temporary = path.with_name('.profile-backup-' + uuid4().hex + '.tmp')
        try:
            temporary.write_text(json.dumps(saved, ensure_ascii=False, indent=2), encoding='utf-8')
            if path.exists():
                backup_temporary.write_bytes(path.read_bytes())
                backup_temporary.replace(path.with_name(path.name + '.bak'))
            temporary.replace(path)
        finally:
            temporary.unlink(missing_ok=True)
            backup_temporary.unlink(missing_ok=True)
    return get_profile(path)
